File: plugins/orchestrator/skills/memory-bank/scripts/task_manager.ts
import { readdirSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { spawn } from 'node:child_process';

interface TimeApiResponse {
  datetime: string;
}

interface Args {
  command: string;
  name?: string;
  request?: string;
  id?: string;
  message?: string;
}

function parseArgs(): Args {
  const args = process.argv.slice(2);
  
  if (args.length === 0) {
    console.log('Usage: bun run task_manager.ts <command> [options]');
    console.log('Commands:');
    console.log('  add <name> [--request <request>]  - Add a new task');
    console.log('  update <id> <message>              - Update a task');
    console.log('  list                               - List all tasks');
    process.exit(1);
  }
  
  const command = args[0];
  const parsed: Args = { command };
  
  if (command === 'add') {
    parsed.name = args[1];
    parsed.request = '';
    
    // Parse --request flag
    const requestIndex = args.indexOf('--request');
    if (requestIndex !== -1 && args[requestIndex + 1]) {
      parsed.request = args[requestIndex + 1];
    }
  } else if (command === 'update') {
    parsed.id = args[1];
    parsed.message = args[2];
  } else if (command !== 'list') {
    console.log(`Unknown command: ${command}`);
    process.exit(1);
  }
  
  return parsed;
}

async function getUtcTime(): Promise<string> {
  try {
    const response = await fetch('http://worldtimeapi.org/api/timezone/UTC', {
      signal: AbortSignal.timeout(5000),
    });
    const data: TimeApiResponse = await response.json();
    return data.datetime;
  } catch {
    return new Date().toISOString();
  }
}

async function main(): Promise<void> {
  const args = parseArgs();
  const rootDir = process.cwd();
  const memoriesDir = join(rootDir, '.memories');
  const scriptDir = dirname(fileURLToPath(import.meta.url));
  
  // Check if .memories directory exists
  const memoriesDirFile = Bun.file(memoriesDir);
  if (!(await memoriesDirFile.exists())) {
    console.log('Error: .memories directory not found. Run init_memories.ts first.');
    return;
  }
  
  if (args.command === 'add') {
    if (!args.name) {
      console.log('Error: Task name is required for add command.');
      process.exit(1);
    }
    
    // Generate ID
    const files = readdirSync(memoriesDir);
    const existingTasks = files.filter((f) => f.startsWith('TASK'));
    const nextId = existingTasks.length + 1;
    const taskId = `TASK${String(nextId).padStart(3, '0')}`;
    
    const utcTime = await getUtcTime();
    
    // Load template
    const templatePath = join(scriptDir, '..', 'assets', 'templates', 'task_template.md');
    const templateFile = Bun.file(templatePath);
    let content = await templateFile.text();
    
    content = content.replace('{{TASK_ID}}', taskId);
    content = content.replace('{{TASK_NAME}}', args.name);
    content = content.replace('{{UTC_DATETIME}}', utcTime);
    content = content.replace('{{DATE}}', utcTime);
    content = content.replace('{{REQUEST}}', args.request ?? '');
    
    const filename = `${taskId}-${args.name.replace(/\s+/g, '-').toLowerCase()}.md`;
    const filepath = join(memoriesDir, filename);
    
    await Bun.write(filepath, content);
    
    // Update index
    const indexPath = join(memoriesDir, '_index.md');
    const indexFile = Bun.file(indexPath);
    if (await indexFile.exists()) {
      const indexContent = await indexFile.text();
      await Bun.write(indexPath, `${indexContent}\n- [${taskId}] ${args.name} - Pending`);
    }
    
    console.log(`Created task ${taskId}: ${filename}`);
  } else if (args.command === 'update') {
    if (!args.id || !args.message) {
      console.log('Error: Task ID and message are required for update command.');
      process.exit(1);
    }
    
    // Find file by ID
    const files = readdirSync(memoriesDir);
    let targetFile: string | null = null;
    
    for (const f of files) {
      if (f.startsWith(args.id!)) {
        targetFile = f;
        break;
      }
    }
    
    if (!targetFile) {
      console.log(`Task ${args.id} not found.`);
      return;
    }
    
    const filepath = join(memoriesDir, targetFile);
    const utcTime = await getUtcTime();
    
    const file = Bun.file(filepath);
    const content = await file.text();
    const updateText = `\n### ${utcTime}\n- ${args.message}\n`;
    
    await Bun.write(filepath, content + updateText);
    
    console.log(`Updated ${targetFile}`);
  } else if (args.command === 'list') {
    console.log('Tasks:');
    const files = readdirSync(memoriesDir);
    for (const f of files) {
      if (f.startsWith('TASK')) {
        console.log(` - ${f}`);
      }
    }
  }
  
  // Auto-sync
  const syncScriptPath = join(scriptDir, 'sync_metadata.ts');
  console.log('\nRunning sync_metadata.ts...');
  
  return new Promise<void>((resolve, reject) => {
    const child = spawn(process.execPath, ['run', syncScriptPath], {
      stdio: 'inherit',
    });
    
    child.on('close', (code) => {
      if (code === 0) {
        resolve();
      } else {
        reject(new Error(`sync_metadata.ts exited with code ${code}`));
      }
    });
    
    child.on('error', reject);
  });
}

main().catch((error) => {
  console.error('Error:', error);
  process.exit(1);
});

File: plugins/orchestrator/skills/memory-bank/scripts/task_manager.py

import sys
import os
import argparse
import datetime
import urllib.request
import json
import re

def get_utc_time():
    try:
        with urllib.request.urlopen("http://worldtimeapi.org/api/timezone/UTC", timeout=5) as response:
            data = json.loads(response.read().decode())
            return data["datetime"]
    except:
        return datetime.datetime.now(datetime.timezone.utc).isoformat()

def main():
    parser = argparse.ArgumentParser(description="Manage Memory Bank tasks")
    subparsers = parser.add_subparsers(dest="command")

    add_parser = subparsers.add_parser("add", help="Add a new task")
    add_parser.add_argument("name", help="Task name")
    add_parser.add_argument("--request", help="Original request", default="")

    update_parser = subparsers.add_parser("update", help="Update a task")
    update_parser.add_argument("id", help="Task ID (e.g., TASK001)")
    update_parser.add_argument("message", help="Progress log message")

    list_parser = subparsers.add_parser("list", help="List tasks")

    args = parser.parse_args()

    root_dir = os.getcwd()
    memories_dir = os.path.join(root_dir, ".memories")
    script_dir = os.path.dirname(os.path.abspath(__file__))
    
    if not os.path.exists(memories_dir):
        print("Error: .memories directory not found. Run init_memories.py first.")
        return

    if args.command == "add":
        # Generate ID
        existing_tasks = [f for f in os.listdir(memories_dir) if f.startswith("TASK")]
        next_id = len(existing_tasks) + 1
        task_id = f"TASK{next_id:03d}"
        
        utc_time = get_utc_time()
        
        # Load template
        template_path = os.path.join(script_dir, "..", "assets", "templates", "task_template.md")
        with open(template_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        content = content.replace("{{TASK_ID}}", task_id)
        content = content.replace("{{TASK_NAME}}", args.name)
        content = content.replace("{{UTC_DATETIME}}", utc_time)
        content = content.replace("{{DATE}}", utc_time) # Simplified
        content = content.replace("{{REQUEST}}", args.request)
        
        filename = f"{task_id}-{args.name.replace(' ', '-').lower()}.md"
        filepath = os.path.join(memories_dir, filename)
        
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        
        # Update index (Simplified append for now)
        index_path = os.path.join(memories_dir, "_index.md")
        if os.path.exists(index_path):
            with open(index_path, 'a', encoding='utf-8') as f:
                f.write(f"\n- [{task_id}] {args.name} - Pending")
        
        print(f"Created task {task_id}: {filename}")

    elif args.command == "update":
        # Find file by ID
        target_file = None
        for f in os.listdir(memories_dir):
            if f.startswith(args.id):
                target_file = f
                break
        
        if not target_file:
            print(f"Task {args.id} not found.")
            return
            
        filepath = os.path.join(memories_dir, target_file)
        utc_time = get_utc_time()
        
        with open(filepath, 'a', encoding='utf-8') as f:
            f.write(f"\n### {utc_time}\n- {args.message}\n")
            
        print(f"Updated {target_file}")

    elif args.command == "list":
         print("Tasks:")
         for f in os.listdir(memories_dir):
            if f.startswith("TASK"):
                print(f" - {f}")

    # Auto-sync
    import subprocess
    sync_script = os.path.join(script_dir, "sync_metadata.py")
    subprocess.run([sys.executable, sync_script])

if __name__ == "__main__":
    main()
